cmd/multitenant-autoscaler/main.go

```diff
@@ -79,12 +84,22 @@ func main() {
 		logger.Fatal("Error building kubernetes clientset.", zap.Error(err))
 	}
 
+	rm := restmapper.NewDeferredDiscoveryRESTMapper(cached.NewMemCacheClient(kubeClientSet.Discovery()))
+	go wait.Until(func() {
```

Comment:
I am uncomfortable with the increasing amount of logic in this method as it is untested/untestable from unit tests and the main method is becoming harder to read because it combines several unrelated sections of code with multiple responsibilities. IMO the purpose of the main method should be dependency wiring and initiation rather than handling detailed contracts to other components. I realise I am probably in the minority here, but I mention this as I think it's a valid, general concern for main methods in the project.

Reply:
I certainly empathize with this, I've spent significant time refactoring the codebase so that we can practically unit test the bulk of our controllers.  However, there are simply aspects of Kubernetes that are impractical to reliably unit test ([see here](https://github.com/kubernetes-sigs/controller-runtime/issues/90)) because you essentially need a real environment to practically test them with reasonable fidelity.  Instead of having testing that spins up an API Server and etcd locally for this, we have our e2e tests.

I think that most of the split between what real logic lives in `./cmd` and `./pkg` today is a function of how reliably we can unit test it (luckily most stuff falls on the right side of this).  I'm certainly receptive to redrawing these lines as we get smarter, or as Kubernetes gets better fakes, but I think those are the limiting factors here.